velox/functions/prestosql/json/JsonStringUtil.cpp

```diff
@@ -0,0 +1,191 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <array>
+
+#include "folly/Unicode.h"
+
+#include "velox/common/base/Exceptions.h"
+#include "velox/functions/prestosql/json/JsonStringUtil.h"
+
+namespace facebook::velox {
+namespace {
+/// Return the length in byte of the next UTF-8 encoded character at the
+/// beginning of `string`. If the beginning of `string` is not valid UTF-8
+/// encoding, return -1.
+FOLLY_ALWAYS_INLINE int validateAndGetNextUtf8Length(
+    const unsigned char* string,
+    const unsigned char* end) {
+  if (string == end) {
+    return 0;
+  }
+  if ((*string & 0x80u) == 0) {
+    return 1;
+  }
+  if ((*string & 0xE0u) == 0xC0u && (string + 1) < end &&
+      (*(string + 1) & 0xC0u) == 0x80u) {
+    return 2;
+  }
+  if ((*string & 0xF0u) == 0xE0u && (string + 2) < end &&
+      (*(string + 1) & 0xC0u) == 0x80u && (*(string + 2) & 0xC0u) == 0x80u) {
+    return 3;
+  }
+  if ((*string & 0xF8u) == 0xF0u && (string + 3) < end &&
+      (*(string + 1) & 0xC0u) == 0x80u && (*(string + 2) & 0xC0u) == 0x80u &&
+      (*(string + 3) & 0xC0u) == 0x80u) {
+    return 4;
+  }
+  return -1;
+}
+
+FOLLY_ALWAYS_INLINE char hexDigit(uint8_t c) {
+  return c < 10 ? c + '0' : c - 10 + 'A';
+}
+
+FOLLY_ALWAYS_INLINE void writeHex(char16_t value, char*& out) {
+  value = folly::Endian::little(value);
+  *out++ = '\\';
+  *out++ = 'u';
+  *out++ = hexDigit((value >> 12) & 0x0F);
+  *out++ = hexDigit((value >> 8) & 0x0F);
+  *out++ = hexDigit((value >> 4) & 0x0F);
+  *out++ = hexDigit(value & 0x0F);
+}
+
+std::array<int8_t, 128> getAsciiEscapes() {
+  std::array<int8_t, 128> escapes;
+  std::fill(escapes.data(), escapes.data() + 32, -1);
+  escapes['"'] = '"';
+  escapes['\\'] = '\\';
+  escapes['\b'] = 'b';
+  escapes['\t'] = 't';
+  escapes['\n'] = 'n';
+  escapes['\f'] = 'f';
+  escapes['\r'] = 'r';
+  return escapes;
+}
+static const std::array<int8_t, 128> asciiEscapes = getAsciiEscapes();
+
+FOLLY_ALWAYS_INLINE void encodeAscii(int8_t value, char*& out) {
+  int8_t escapeCode = asciiEscapes[value];
+  if (escapeCode == 0) {
+    *out++ = char(value);
+  } else if (escapeCode > 0) {
+    *out++ = '\\';
+    *out++ = char(escapeCode);
+  } else {
+    writeHex(value, out);
+  }
+}
+
+FOLLY_ALWAYS_INLINE size_t encodedAsciiSize(uint8_t value) {
+  int8_t escapeCode = asciiEscapes[value];
+  if (escapeCode == 0) {
+    return 1;
+  }
+  if (escapeCode > 0) {
+    return 2;
+  }
+  return 6;
+}
+
+FOLLY_ALWAYS_INLINE void encodeUtf16Hex(char32_t codePoint, char*& out) {
```

Comment:
nit can you add some comments on these magic values and why we are doing what we are doing here ?

Reply:
More comments are added.